tests/integration/services/test_provider.py

```diff
@@ -0,0 +1,52 @@
+# Copyright 2024 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+"""Integration tests for the provider service."""
+
+import pathlib
+import sys
+
+import pytest
+from craft_application.models import BuildInfo
+from craft_cli.pytest_plugin import RecordingEmitter
+
+from charmcraft import services
+from charmcraft.services.provider import _maybe_lock_cache
+
+
+@pytest.mark.skipif(sys.platform == "win32", reason="no cache on windows")
+def test_locks_cache(
+    service_factory: services.CharmcraftServiceFactory,
+    tmp_path: pathlib.Path,
+    default_build_info: BuildInfo,
+    emitter: RecordingEmitter,
+):
+    _maybe_lock_cache(tmp_path)
+    assert (tmp_path / "charmcraft.lock").exists()
+    provider = service_factory.provider
+    provider_kwargs = {
+        "build_info": default_build_info,
+        "work_dir": pathlib.Path(__file__).parent,
+        "cache_path": tmp_path,
+    }
+
+    with provider.instance(**provider_kwargs) as instance:
+        # Because we've already locked the cache, we shouldn't see the lockfile.
```

Comment:
I'm confused here - why isn't the cache visible? Did the call to `_maybe_lock_cache()` inside `provider.instance()` fall into the `except OSError:` codepath?

Reply:
Yes, because we run `_maybe_lock_cache` on line 36. Looking back I realise I named the function for the success case and tested for the failure case. I'll fix that

Reply:
this is surprising to me because I thought `flock()` was per-process, so a process trying to flock a file it already holds would be a noop
If this isn't the case, what happens if we do multiple consecutive managed builds? Like a build plan with multiple entries

Reply:
You're right! I wasn't checking for that properly, either. However, it's not quite per-process either. It's per file descriptor, so the same process can't lock the same path in two places if it's got the file open separately twice. (This also requires making two separate `Path` objects, as `pathlib` will cache the file descriptor for the same object, as I learnt the hard way.)

I've updated it now with a much better lock and test.